rails/database-migrations.md

```diff
@@ -0,0 +1,164 @@
+# Database Migration Best Practices
+
+## Principles
+
+### Use the strong_migrations library
+https://github.com/ankane/strong_migrations helps catch many common database migration mistakes, especially when using Postgres.
+
+### Use Rails migrations for simple data migrations
+Using Rails' `db:migrate` to migrate data has many advantages over custom scripts:
+
+- Rails schema versioning makes it easy to know the status of each migration.
+- It's simple for other developers on your team. They only need to `git pull` and then run `db:migrate`.
+- Deployment pipelines are usually set up to run data migrations, so there are no extra steps to run.
+
+Of course, there are exceptions to this rule. Super complex data migrations may need more involved scripts.
+
+### Keep migrations runnable for as long as possible
+Migrations that break after only a few git commits are a nightmare for teams. This problem is almost always caused by migration code relying on application code like model associations and methods. A migration runs at a single point in time, whereas application code is always changing, so the two get out of sync.
+
+For example, suppose a migration is added to the master branch which uses an association defined on an ActiveRecord model. Two commits later, the association is removed from the ActiveRecord model, but the migration code isn't updated.
+
+Now, another developer rebases his feature branch onto master and attempts to run the pending migration.
+
+Failure. To be specific, a `NoMethodError: undefined method` failure because the association the migration depended on isn't defined on the model anymore, so the migration blows up. 😩
+
+You can avoid this problem by following a simple rule:
+
+> Migrations may only use code defined within their own class.
+
+If you're migrating data and need helper methods or ActiveRecord associations, define them directly in the migration class itself. See the **Use Cases** section below for an example of how to properly use ActiveRecord associations in a migration.
+
+
+### Do not use migrations to create new data
+Use `seeds.rb` or a rake task to populate necessary app data. Otherwise running `db:schema:load` will leave the app's database without the needed data.
+
+### Always test your migrations
+Always run the migration, roll it back, and re-run it (`rails db:migrate:redo` is handy for this). This prevents accidentally writing irreversible migrations because you forgot to specify all the needed details the `change` method needs to calculate the reverse of the operation you're running.
+
+If a migration by its nature cannot be rolled back, be explicit about it by raising `IrreversibleMigration` in the `down` method.
```

Comment:
I'd define a cutout here for a slightly different philosophy. Historically at McKesson and other projects, I've written migrations that did data cleanup or transformation, that were technically irreversible but actually didn't matter. So as not to constrain up/down migrations, we codified writing a migration with a `#down` that would say "This migration is not reversible, but that's probably okay" and then pass onwards.